src/FilenameParsing/FileIDHelperResolutionStrategy.php

```diff
@@ -0,0 +1,334 @@
+<?php
+
+namespace SilverStripe\Assets\FilenameParsing;
+
+use League\Flysystem\Filesystem;
+use SilverStripe\Versioned\Versioned;
+use SilverStripe\Assets\File;
+use SilverStripe\ORM\DB;
+
+/**
+ * File resolution strategy that relies on a list of FileIDHelpers to find files.
+ * * `DefaultFileIDHelper` is the default helper use to generate new file ID.
+ * * `ResolutionFileIDHelpers` can contain a list of helpers that will be used to try to find existing file.
+ *
+ * This file resolution strategy can be helpfull when the approach to resolving files has changed over time and you need
+ * older file format to resolve.
+ *
+ * You may also provide a `VersionedStage` to only look at files that were published.
+ */
+class FileIDHelperResolutionStrategy implements FileResolutionStrategy
+{
+    /**
+     * The FileID helper that will be use to build FileID for this adapter.
+     * @var FileIDHelper
+     */
+    private $defaultFileIDHelper;
+
+    /**
+     * List of FileIDHelper that should be use to try to parse FileIDs on this adapter.
+     * @var FileIDHelper[]
+     */
+    private $resolutionFileIDHelpers;
+
+    /**
+     * Constrain this strategy to the a specific versioned stage.
+     * @var string
+     */
+    private $versionedStage = Versioned::DRAFT;
+
+    public function resolveFileID($fileID, Filesystem $filesystem)
+    {
+        foreach ($this->resolutionFileIDHelpers as $fileIDHelper) {
+            $parsedFileID = $fileIDHelper->parseFileID($fileID);
+            if (!$parsedFileID) {
+                continue;
+            }
+
+            if ($redirect = $this->searchForTuple($parsedFileID, $filesystem, false)) {
```

Comment:
That naming is a bit confusing - it's not actually a redirect in the HTTP sense, but rather a different ParsedFileID, right?

Reply:
That was a left over from my SS4.3 redirect fix. Restructured things a bit so it's clearer.